c/tskit/trees.c

```diff
@@ -2076,19 +2076,661 @@ typedef struct {
+        alt_allele = site->ancestral_state;
+        alt_allele_length = site->ancestral_state_length;
+        if (mutation.parent != TSK_NULL) {
+            parent_mut = site->mutations[mutation.parent - site->mutations[0].id];
+            alt_allele = parent_mut.derived_state;
+            alt_allele_length = parent_mut.derived_state_length;
+        }
+        for (allele = 0; allele < num_alleles; allele++) {
+            if (alt_allele_length == allele_lengths[allele]
+                && tsk_memcmp(alt_allele, alleles[allele], allele_lengths[allele])
+                       == 0) {
+                break;
+            }
+        }
+        tsk_bug_assert(allele < num_alleles);
+
+        tsk_bit_array_get_row(out_allele_samples, allele, &alt_allele_samples_row);
+        tsk_bit_array_subtract(&alt_allele_samples_row, &allele_samples_row);
+    }
+    *out_num_alleles = num_alleles;
+out:
+    tsk_safe_free(alleles);
+    tsk_safe_free(allele_lengths);
+    return ret;
+}
+
+static int
+get_mutation_samples(const tsk_treeseq_t *self, const tsk_size_t tree_index,
+    const tsk_id_t *right_child, const tsk_id_t *left_sib, const tsk_id_t *parent,
+    const tsk_size_t *site_offsets, tsk_bit_array_t *allele_samples,
+    tsk_size_t *restrict *num_alleles)
+// TODO: the restricts here, not sure. Are the row declarations aliases? I think so.
+{
+    int ret = 0;
+
+    const tsk_size_t num_edges = tsk_treeseq_get_num_edges(self);
+    const tsk_size_t num_samples = tsk_treeseq_get_num_samples(self);
+    const tsk_size_t num_sites = self->tree_sites_length[tree_index];
+    const tsk_flags_t *restrict flags = self->tables->nodes.flags;
+    const tsk_site_t *restrict tree_sites = self->tree_sites[tree_index];
+    const tsk_id_t *restrict mut_nodes = self->tables->mutations.node;
+    const tsk_site_t *restrict site;
+
+    tsk_id_t u, node, top_mut_node, *stack = NULL;
+    tsk_bit_array_t mut_samples, mut_samples_row, node_paths, path, out_row;
+    tsk_size_t num_mutations, mut_samples_offset, mut_offset;
+    tsk_size_t s, m;
+    int stack_top;
+
+    // Get the number of mutations and the top mutation node
+    num_mutations = 0;
+    top_mut_node = 0;
+    for (s = 0; s < num_sites; s++) {
+        site = &tree_sites[s];
+        for (m = 0; m < site->mutations_length; m++) {
+            num_mutations++;
+            if (site->mutations[m].node > top_mut_node) {
+                top_mut_node = site->mutations[m].node;
+            }
+        }
+    }
+
+    ret = tsk_bit_array_init(&mut_samples, num_samples, num_mutations);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_bit_array_init(
+        &node_paths, tsk_treeseq_get_num_nodes(self), num_mutations);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Size of stack is derived from tsk_tree_get_size_bound
+    stack = tsk_malloc((1 + self->num_samples + num_edges) * sizeof(*stack));
```

Comment:
This is actually a substantial over-estimate of the  max stack size - the simplest thing to do there is to the ``ts.num_nodes`` as the maximum stack size (when we don't know anything about the individual trees that it'll be used on)